File: submission/241098117-张城宗/src/core algorithm/AI/ai_engine.py
import sys, os
from typing import Optional, List, Tuple

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from triangular_game import TriangularGame, Player, PointState


class MinimaxAI:
    def __init__(self, depth: int = 3):
        self.depth = depth

    def get_legal_moves(self, game, player: Player) -> List[Tuple[int, int]]:
        return [pos for pos in game.grid if game._is_legal_move(pos, player)]

    def order_moves(
        self,
        game,
        moves: List[Tuple[int, int]],
        player: Player,
    ) -> List[Tuple[int, int]]:
        opponent_line = (
            PointState.WHITE_LINE if player == Player.BLACK else PointState.BLACK_LINE
        )
        own_line = (
            PointState.BLACK_LINE if player == Player.BLACK else PointState.WHITE_LINE
        )
        ordered_moves = [[], [], []]

        for pos in moves:
            if game.grid[pos] == opponent_line:
                ordered_moves[0].append(pos)
            elif any(
                game.grid[adjacent] == own_line
                for adjacent in game._get_adjacent_positions(pos)
            ):
                ordered_moves[1].append(pos)
            else:
                ordered_moves[2].append(pos)

        return (ordered_moves[0] + ordered_moves[1] + ordered_moves[2])[:20]

    def minimax(
        self,
        game,
        depth: int,
        alpha: float,
        beta: float,
        maximizing_player: bool,
        ai_player: Player,
    ) -> float:
        if game.game_over:
            return game.evaluate(ai_player)
        if depth == 0:
            return game.evaluate(ai_player)

        legal_moves = self.get_legal_moves(game, game.current_player)
        if not legal_moves:
            snapshot = game._save_state()
            game.consecutive_skips += 1
            if game.consecutive_skips >= 2:
                game.game_over = True
            else:
                game._switch_player()
            score = self.minimax(
                game, depth - 1, alpha, beta, not maximizing_player, ai_player
            )
            game._restore_state(snapshot)
            return score

        ordered = self.order_moves(game, legal_moves, game.current_player)
        if maximizing_player:
            best_val = float("-inf")
            for pos in ordered:
                snapshot = game._save_state()
                game._apply_move_for_ai(pos)
                val = self.minimax(game, depth - 1, alpha, beta, False, ai_player)
                game._restore_state(snapshot)
                best_val = max(best_val, val)
                alpha = max(alpha, best_val)
                if beta <= alpha:
                    break
            return best_val

        best_val = float("inf")
        for pos in ordered:
            snapshot = game._save_state()
            game._apply_move_for_ai(pos)
            val = self.minimax(game, depth - 1, alpha, beta, True, ai_player)
            game._restore_state(snapshot)
            best_val = min(best_val, val)
            beta = min(beta, best_val)
            if beta <= alpha:
                break
        return best_val

    def get_best_move(
        self, game, ai_player: Player
    ) -> Optional[Tuple[int, int]]:
        legal_moves = self.get_legal_moves(game, ai_player)
        if not legal_moves:
            return None

        ordered = self.order_moves(game, legal_moves, ai_player)
        best_move = None
        best_val = float("-inf")
        for pos in ordered:
            snapshot = game._save_state()
            game._apply_move_for_ai(pos)
            val = self.minimax(
                game,
                self.depth - 1,
                float("-inf"),
                float("inf"),
                False,
                ai_player,
            )
            game._restore_state(snapshot)
            if val > best_val:
                best_val = val
                best_move = pos
        return best_move

    def get_top_moves(
        self, game, ai_player: Player, top_n: int = 5
    ) -> List[Tuple[Tuple[int, int], float]]:
        legal_moves = self.get_legal_moves(game, ai_player)
        if not legal_moves:
            return []

        ordered = self.order_moves(game, legal_moves, ai_player)
        scored_moves = []
        for pos in ordered:
            snapshot = game._save_state()
            game._apply_move_for_ai(pos)
            score = self.minimax(
                game,
                self.depth - 1,
                float("-inf"),
                float("inf"),
                False,
                ai_player,
            )
            game._restore_state(snapshot)
            scored_moves.append((pos, score))

        scored_moves.sort(key=lambda item: item[1], reverse=True)
        return scored_moves[:top_n]

File: submission/241098117-张城宗/src/web/web前端/AIEngine.js
import { GameEngine, Player, PointState } from "./GameEngine.js";

function pointKey(point) {
  return `${point[0]},${point[1]}`;
}

function getOpponents(engine, player) {
  return engine._getOpponents
    ? engine._getOpponents(player)
    : engine.activePlayers.filter((candidate) => candidate !== player);
}

function cloneCachedTerritories(engine) {
  return Object.fromEntries(
    engine.activePlayers.map((player) => {
      const territory = engine.cachedTerritories[player];
      return [
        player,
        {
          ...territory,
          polygon: territory.polygon
            ? territory.polygon.map((point) => [...point])
            : null,
        },
      ];
    }),
  );
}

function saveState(engine) {
  return {
    grid: new Map(engine.grid),
    edges: {
      BLACK: new Set(engine.edges.BLACK ?? []),
      WHITE: new Set(engine.edges.WHITE ?? []),
      PURPLE: new Set(engine.edges.PURPLE ?? []),
    },
    historyHashes: new Set(engine.historyHashes),
    consecutiveSkips: engine.consecutiveSkips,
    currentPlayer: engine.currentPlayer,
    gameOver: engine.gameOver,
    turnCount: engine.turnCount,
    cachedTerritories: cloneCachedTerritories(engine),
  };
}

function restoreState(engine, snapshot) {
  engine.grid = new Map(snapshot.grid);
  engine.edges = {
    BLACK: new Set(snapshot.edges.BLACK),
    WHITE: new Set(snapshot.edges.WHITE),
    PURPLE: new Set(snapshot.edges.PURPLE),
  };
  engine.historyHashes = new Set(snapshot.historyHashes);
  engine.consecutiveSkips = snapshot.consecutiveSkips;
  engine.currentPlayer = snapshot.currentPlayer;
  engine.gameOver = snapshot.gameOver;
  engine.turnCount = snapshot.turnCount;
  engine.cachedTerritories = snapshot.cachedTerritories;
}

function applyMoveForAI(engine, point) {
  const result = engine.playMove(point);
  if (!result.success) {
    const reason = String(result.reason ?? "").toUpperCase();
    if (reason.includes("SUPERKO") || reason.includes("REPEAT")) {
      return false;
    }
    return false;
  }
  return true;
}

function fastTerritoryBFS(engine, player) {
  const myStates = engine._getPlayerStates(player);
  const opponentStates = new Set();
  for (const opponent of getOpponents(engine, player)) {
    const states = engine._getPlayerStates(opponent);
    opponentStates.add(states.node);
    opponentStates.add(states.line);
  }

  const frontier = [];
  const visited = new Set();
  for (const point of engine.validPositions) {
    const state = engine._getState(point);
    if (state === myStates.node || state === myStates.line) {
      const key = pointKey(point);
      visited.add(key);
      frontier.push(point);
    }
  }

  for (let index = 0; index < frontier.length; index += 1) {
    for (const adjacent of engine.getAdjacentPositions(frontier[index])) {
      const key = pointKey(adjacent);
      if (!visited.has(key) && !opponentStates.has(engine._getState(adjacent))) {
        visited.add(key);
        frontier.push(adjacent);
      }
    }
  }

  return visited.size;
}

function evaluate(engine, player) {
  const opponents = getOpponents(engine, player);
  const myNodes = engine._getPlayerNodes(player);
  const myLines = engine._getPlayerLines(player);

  let opponentNodeCount = 0;
  let opponentLineCount = 0;
  const opponentLines = [];
  let mainOpponent = null;
  let mainOpponentNodeCount = -1;

  for (const opponent of opponents) {
    const nodes = engine._getPlayerNodes(opponent);
    const lines = engine._getPlayerLines(opponent);
    opponentNodeCount += nodes.length;
    opponentLineCount += lines.length;
    opponentLines.push(...lines);

    if (nodes.length > mainOpponentNodeCount) {
      mainOpponent = opponent;
      mainOpponentNodeCount = nodes.length;
    }
  }

  const nodeAdvantage = myNodes.length - opponentNodeCount;
  const coverageAdvantage =
    myNodes.length + myLines.length - opponentNodeCount - opponentLineCount;
  const territoryAdvantage =
    fastTerritoryBFS(engine, player) -
    (mainOpponent === null ? 0 : fastTerritoryBFS(engine, mainOpponent));

  const legalMoveKeys = new Set(
    engine.getLegalMoves(player).map((move) => pointKey(move.point)),
  );
  const attackThreats = opponentLines.reduce(
    (count, point) => count + (legalMoveKeys.has(pointKey(point)) ? 1 : 0),
    0,
  );
  const connectionQuality = myNodes.reduce(
    (score, node) => score + (engine._isConnectedToInitial(node, player) ? 1 : -1),
    0,
  );

  return (
    15 * nodeAdvantage +
    8 * coverageAdvantage +
    20 * territoryAdvantage +
    12 * attackThreats +
    10 * connectionQuality
  );
}

function orderMoves(engine, moves, player) {
  const ownLine = engine._getPlayerStates(player).line;
  const opponentLines = new Set(
    getOpponents(engine, player).map(
      (opponent) => engine._getPlayerStates(opponent).line,
    ),
  );
  const tiers = [[], [], []];

  for (const point of moves) {
    if (opponentLines.has(engine._getState(point))) {
      tiers[0].push(point);
    } else if (
      engine
        .getAdjacentPositions(point)
        .some((adjacent) => engine._getState(adjacent) === ownLine)
    ) {
      tiers[1].push(point);
    } else {
      tiers[2].push(point);
    }
  }

  return [...tiers[0], ...tiers[1], ...tiers[2]].slice(0, 20);
}

class MinimaxAI {
  constructor(depth = 3) {
    this.depth = depth;
  }

  getLegalMoves(engine, player) {
    return engine.getLegalMoves(player).map((move) => move.point);
  }

  orderMoves(engine, moves, player) {
    return orderMoves(engine, moves, player);
  }

  minimax(engine, depth, alpha, beta, maximizingPlayer, aiPlayer) {
    if (engine.gameOver || depth === 0) {
      return evaluate(engine, aiPlayer);
    }

    const legalMoves = this.getLegalMoves(engine, engine.currentPlayer);
    if (legalMoves.length === 0) {
      const snapshot = saveState(engine);
      engine.consecutiveSkips += 1;
      if (engine.consecutiveSkips >= engine.activePlayers.length) {
        engine.gameOver = true;
      } else {
        engine._switchPlayer();
      }
      const score = this.minimax(
        engine,
        depth - 1,
        alpha,
        beta,
        !maximizingPlayer,
        aiPlayer,
      );
      restoreState(engine, snapshot);
      return score;
    }

    const ordered = this.orderMoves(engine, legalMoves, engine.currentPlayer);
    if (maximizingPlayer) {
      let bestValue = Number.NEGATIVE_INFINITY;
      for (const point of ordered) {
        const snapshot = saveState(engine);
        const applied = applyMoveForAI(engine, point);
        if (applied) {
          const value = this.minimax(engine, depth - 1, alpha, beta, false, aiPlayer);
          bestValue = Math.max(bestValue, value);
          alpha = Math.max(alpha, bestValue);
        }
        restoreState(engine, snapshot);
        if (beta <= alpha) {
          break;
        }
      }
      return bestValue;
    }

    let bestValue = Number.POSITIVE_INFINITY;
    for (const point of ordered) {
      const snapshot = saveState(engine);
      const applied = applyMoveForAI(engine, point);
      if (applied) {
        const value = this.minimax(engine, depth - 1, alpha, beta, true, aiPlayer);
        bestValue = Math.min(bestValue, value);
        beta = Math.min(beta, bestValue);
      }
      restoreState(engine, snapshot);
      if (beta <= alpha) {
        break;
      }
    }
    return bestValue;
  }

  getTopMoves(engine, aiPlayer, topN = 5) {
    const legalMoves = this.getLegalMoves(engine, aiPlayer);
    if (legalMoves.length === 0) {
      return [];
    }

    const ordered = this.orderMoves(engine, legalMoves, aiPlayer);
    const scoredMoves = [];
    for (const point of ordered) {
      const snapshot = saveState(engine);
      const applied = applyMoveForAI(engine, point);
      if (applied) {
        const score = this.minimax(
          engine,
          this.depth - 1,
          Number.NEGATIVE_INFINITY,
          Number.POSITIVE_INFINITY,
          false,
          aiPlayer,
        );
        scoredMoves.push({ point, score });
      }
      restoreState(engine, snapshot);
    }

    scoredMoves.sort((a, b) => b.score - a.score);
    return scoredMoves.slice(0, topN);
  }
}

export { MinimaxAI, evaluate, fastTerritoryBFS, saveState, restoreState };
